include/xtensor-zarr/xzarr_hierarchy.hpp

```diff
@@ -0,0 +1,157 @@
+/***************************************************************************
+* Copyright (c) Wolf Vollprecht, Sylvain Corlay and Johan Mabille          *
+* Copyright (c) QuantStack                                                 *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_ZARR_XZARR_HIERARCHY_HPP
+#define XTENSOR_ZARR_XZARR_HIERARCHY_HPP
+
+#include <iomanip>
+#include <fstream>
+#include <iostream>
+
+#include "nlohmann/json.hpp"
+#include "thirdparty/filesystem.hpp"
+#include "xtensor/xchunked_array.hpp"
+#include "xtensor/xchunk_store_manager.hpp"
+#include "xtensor/xfile_array.hpp"
+#include "xtensor/xdisk_io_handler.hpp"
+#include "xtensor-io/xblosc.hpp"
+#include "xzarr_attrs.hpp"
+
+namespace fs = ghc::filesystem;
+
+namespace xt
+{
+
+    fs::path get_meta_path(fs::path& hier_path, const char* array_path)
+    {
+        while (*array_path == '/')
+        {
+            array_path++;
+        }
+        auto meta_path = hier_path / "meta" / "root" / array_path;
+        fs::create_directories(meta_path.parent_path());
+        return meta_path;
+    }
+
+    fs::path get_data_path(fs::path& hier_path, const char* array_path)
+    {
+        while (*array_path == '/')
+        {
+            array_path++;
+        }
+        auto data_path = hier_path / "data" / array_path;
+        fs::create_directories(data_path);
+        return data_path;
+    }
+
+    class xzarr_hierarchy
+    {
+    public:
+        xzarr_hierarchy(const char* path): m_path(path) {};
+
+        void create_hierarchy();
+
+        template <class value_type, class shape_type, class io_handler = xdisk_io_handler<xblosc_config>>
+        auto create_array(const char* path, shape_type shape, shape_type chunk_shape, nlohmann::json& attrs=nlohmann::json::array());
+
+        template <class value_type, class io_handler = xdisk_io_handler<xblosc_config>>
+        auto get_array(const char* path);
+
```

Comment:
It could be nice to have a `tensor_type` inner type and return it instead of specifying auto as the return type.

Reply:
The problem is that this `tensor_type` cannot be declared at the class level, since a zarr hierarchy can store arrays of different types.

Reply:
the tensor type can be a template:

```cpp
template <class T, class IO>
using tensor_type =  xchunked_array<xchunk_store_manager<xfile_array<T, IO>>, xzarr_attrs>;

// ....

template <class value_type, class io_handler = xdisk_io_handler<xblosc_config>>
tensor_type<value_type, IO_handler> get_array(const char* path);
```

Reply:
:+1: